src/Simulation/QCTraceSimulator/SortedQubitPool.cs

```diff
@@ -0,0 +1,321 @@
+                if (result > 0) {
+                    if (a == Sample && b != Sample) {
+                        MaxLowerBound = ComplexTime.Max(MaxLowerBound, b.Time);
+                    } else if (a != Sample && b == Sample) {
+                        MinUpperBound = ComplexTime.Min(MinUpperBound, a.Time);
+                    }
+                } else if (result < 0) {
+                    if (a == Sample && b != Sample) {
+                        MinUpperBound = ComplexTime.Min(MinUpperBound, b.Time);
+                    } else if (a != Sample && b == Sample) {
+                        MaxLowerBound = ComplexTime.Max(MaxLowerBound, a.Time);
+                    }
+                } else {
+                    // We found the value if one argument is the sample and nother is not.
+                    if ((a == Sample) != (b == Sample)) {
+                        MaxLowerBound = b.Time;
+                        MinUpperBound = b.Time;
+                    }
+                }
+                return result;
+            }
+        }
+
+        /// <summary>
+        /// QubitTimeNodes sorted by Time field. Nodes with the same Time field are stored in a linked list.
+        /// </summary>
+        private readonly SortedSet<QubitTimeNode> QubitTime;
+
+        /// <summary>
+        /// Comparer of two nodes, which compares them by Time field and collects additional data.
+        /// </summary>
+        private readonly VisitingComparer NodeComparer;
+
+        /// <summary>
+        /// Sample which is been compared to the nodes stored in the set.
+        /// Multiple comparisons are done by SortedSet to find if this Sample is present in the set.
+        /// </summary>
+        private readonly QubitTimeNode Sample = new QubitTimeNode(); // We reuse same object to avoid allocations.
+
+        public SortedQubitPool() {
+            NodeComparer = new VisitingComparer();
+            QubitTime = new SortedSet<QubitTimeNode>(NodeComparer);
+        }
+
+        /// <summary>
+        /// Add Qubit to the set with specified time. Multiple qubits with the same time can be added.
+        /// Complexity is log(N), where N - number of nodes.
+        /// </summary>
+        /// <param name="qubitId">Id of a qubit to add to the set.</param>
+        /// <param name="time">Time of the qubit.</param>
+        public void Add(long qubitId, ComplexTime time) {
+            QubitTimeNode newNode = new QubitTimeNode(qubitId, time);
+            if (QubitTime.Add(newNode)) {
+                // New item was added to the set, we are done.
+                return;
+            }
+            if (!QubitTime.TryGetValue(newNode, out QubitTimeNode existingNode)) {
+                // We cannot add, but we cannot find a node with the same value.
+                // Is it a floating point glitch? We'll just ignore this one.
+                // If this happens reuse will be hindered, but program will still work.
```

Comment:
Assert? How would you notice that this is happening (if it does) and reuse isn't as efficient as it could be?

Reply:
Minimum width is not guaranteed anywhere, it's all based on heuristics. So the program will continue correctly with less reuse.
I believe we discourage debug asserts in runtime code. Added Debug.Assert nevertheless.